# 1) About

## Purpose
This notebook transforms raw transaction data into a **temporal behavioral dataset** suitable for Machine Learning Modeling. Each row in final dataset represents one client at one monthly snapshot, with aggregated behavioral features and a **forward looking** fraud label.


## Connection to EDA
In the EDA Notebook, we established:

- Dataset contains 2 AML patterns. Fan-in (multiple senders -> one receiver) and Cycle (circular flows).
- Only ~0.13% of the transactions are fraudulent -> dataset heavily imbalanced.
- Fraudulent clients demonstrate distinct behavioral signals -> 2x unique senders, 2x received trx volume, ~60% more transactions.
- Transaction network is sparse (1.3% density), which makes 'counterparty count' related features discriminative.

## Outcome of the NB
This notebook will transform raw csv files (accounts/transactions/alerts) into a client-month level dataset ready for train-test split.

## Feature Categories
1. Behavioral Intensity: Trx counts/volumes
2. Fan-in Signals: Unique senders, sender diversity, received trx volume
3. Cycle Signals: Sender-receiver overlap, bidirectional behavior patterns
4. Velocity/Change: Activity spikes, new counterparties.

# 2) Imports and Data Loading

In [1]:
import pandas as pd, numpy as np 
from pathlib import Path 
import matplotlib.pyplot as plt 
import seaborn as sns

sns.set_style('whitegrid')
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
pd.set_option('display.expand_frame_repr', False)
pd.set_option("display.float_format", "{:,.2f}".format)

# import raw data
data_path = Path("../data/raw")
dfs = {}
for f in data_path.glob("*.csv"):
    dfs[f.stem] = pd.read_csv(f)

# handling timestamp - assuming 2020-01-01 as the start date of the timestamp
start_date = pd.to_datetime("2020-01-01")
for i, df in dfs.items():
    if 'TIMESTAMP' in df.columns:
        df['date'] = start_date + pd.to_timedelta(df['TIMESTAMP'], unit='D')

df_accounts = dfs['accounts']
df_transactions = dfs['transactions']
df_alerts = dfs['alerts']   

# Tagging fraudulent accounts
fraudulent_accounts = list(df_accounts[df_accounts['IS_FRAUD'] == 1]['ACCOUNT_ID'].unique())

# General Overview
print(f"Accounts: {len(df_accounts):,}, Fraudulent accounts: {len(fraudulent_accounts)}")
print(f"All Transactions: {len(df_transactions):,}, Alerted Transactions: {len(df_alerts):,}")
print(f"Date range of trx: {df_transactions['date'].min().date()} to {df_transactions['date'].max().date()}")


Accounts: 10,000, Fraudulent accounts: 1685
All Transactions: 1,323,234, Alerted Transactions: 1,719
Date range of trx: 2020-01-01 to 2020-07-18


# 3) Snapshot Calendar

Let's define monthly snapshot dates. Each one represents a "prediction point" where we ask: **will this client exhibit suspicious behavior this month?**

**Design rules:**
- Features use **only past data** (up to snapshot date ->exclusive)
- Labels look at the **full calendar month ahead** (snapshot month)
- Need at least **3 full calendar months of history** for rolling features
- Trx data range: 2020-01-01 to 2020-07-18

This means:
- **First usable snapshot:** April 1 (Jan, Feb, Mar -> 3 full months of history)
- **Last usable snapshot:** June 1 (needs full June for labels -> July only has 18 days)

In [2]:
data_start = df_transactions['date'].min()
data_end = df_transactions['date'].max()
print(f"Data covers a period of {(data_end - data_start).days} days")
print(f"Date Range of Transactions: {data_start} to {data_end}")

# Create Monthly Snapshots
# min history - 3 full calendar months
min_history_months = 3

# all month starts
month_starts = pd.date_range(start=data_start, end=data_end, freq = "MS")

# snapshot_dates -> 3 calendar months of history, 1 calendar month of future
snapshot_dates = []
for i in month_starts:

    # does it have full data for this month?
    month_end = i + pd.offsets.MonthEnd(0)
    has_full_month_ahead = month_end <= data_end

    # does it have enough history? 3M 
    history_start = i - pd.DateOffset(months = min_history_months)
    has_enough_history = history_start >= data_start

    if has_full_month_ahead and has_enough_history:
        snapshot_dates.append(i)

print(f"\nUsable snapshot dates ({len(snapshot_dates)}):")

# print each snapshot date's rolling window history start and prediction date
for i in snapshot_dates:
    history_start = i - pd.DateOffset(months = min_history_months) 
    history_end =  i - pd.Timedelta(days=1)
    prediction_end = i + pd.offsets.MonthEnd(0)
        # print(f"SnapshotDate {i}, uses history from {history_start} to i - , {month_end}")
    print(f" {i.date()}  |  history from: {history_start.date()} to {history_end.date()} |  predicts: {i.date()} until {prediction_end.date()}")


Data covers a period of 199 days
Date Range of Transactions: 2020-01-01 00:00:00 to 2020-07-18 00:00:00

Usable snapshot dates (3):
 2020-04-01  |  history from: 2020-01-01 to 2020-03-31 |  predicts: 2020-04-01 until 2020-04-30
 2020-05-01  |  history from: 2020-02-01 to 2020-04-30 |  predicts: 2020-05-01 until 2020-05-31
 2020-06-01  |  history from: 2020-03-01 to 2020-05-31 |  predicts: 2020-06-01 until 2020-06-30


# 4) Feature Engineering

For each client at each snapshot date, we will compute behavioral futures using transactions that occured before the snapshot date.  



## 4.a Behavioral Intensity

General transaction behavior per client, computed for both **Recent** (previous 1 month) and **Baseline** (2 months before that) windows. 

Windows are non-overlapping so velocity ratios give clean change signals.

Features per window:
- Transaction counts, volumes (sum & mean)

In [3]:
def compute_intensity_features(df_transactions, snapshot, window_start, prefix):
    """
    Compute behavioral intensity features for all accounts within a time window - 
    prefix: baseline or recent
    Returns a df with one row per account - per
    """

    # Filter trx per window of interest (last month or 2 months before last month)
    window_trx = df_transactions[(df_transactions["date"] >= window_start) & (df_transactions["date"] < snapshot)]

    # 1. Sending behavior per client
    sent = (
        window_trx.groupby("SENDER_ACCOUNT_ID")
        .agg(
            trx_sent=("TX_ID", "nunique"),
            amt_sent=("TX_AMOUNT", "sum"),
            avg_amt_sent=("TX_AMOUNT", "mean"),
        )
        .rename_axis("ACCOUNT_ID")
    )

    # 2. Receiving behavior per client
    received = (
        window_trx.groupby("RECEIVER_ACCOUNT_ID")
        .agg(
            trx_received=("TX_ID", "nunique"),
            amt_received=("TX_AMOUNT", "sum"),
            avg_amt_received=("TX_AMOUNT", "mean"),
        )
        .rename_axis("ACCOUNT_ID")
    )

    # Combine features
    features = sent.join(received, how="outer").fillna(0)
    # #rename cols
    features.columns =[f"{prefix}_{col}" for col in features.columns]
    # features.columns =[col if col == "ACCOUNT_ID" else f"{prefix}_{col}" for col in features.columns]
    return features

# Test: compute recent features for April 1 snapshot
test_snapshot = snapshot_dates[0]
test_recent_start = test_snapshot - pd.DateOffset(months=1)

recent_features = compute_intensity_features(df_transactions, test_snapshot, test_recent_start, "recent")
print(f"Snapshot: {test_snapshot.date()}")
print(f"Window: {test_recent_start.date()} to {test_snapshot.date()}")
print(f"Accounts with activity: {len(recent_features)}")
print(f"\nSample (first 5 accounts):")
recent_features.head()


Snapshot: 2020-04-01
Window: 2020-03-01 to 2020-04-01
Accounts with activity: 9995

Sample (first 5 accounts):


,recent_trx_sent,recent_amt_sent,recent_avg_amt_sent,recent_trx_received,recent_amt_received,recent_avg_amt_received
ACCOUNT_ID,,,,,,
1,3.00,527.40,175.80,0.00,0.00,0.00
2,2.00,284.10,142.05,0.00,0.00,0.00
3,2.00,251.76,125.88,3.00,40.89,13.63
4,5.00,755.65,151.13,0.00,0.00,0.00
5,2.00,140.48,70.24,2.00,16.50,8.25


## 4.b Fan-In Signals

Features that capture fan-in signals -> multiple senders funneling money to a single receiver. 

From the EDA, we know that fraudulent clients have 2x as many unique senders and 2x received volume compared to non-fraud clients.

Features:

- Unique Senders: Number of unique accounts sending funds to this account
- Received Volume Concentration: What % of received trx comes from the top sender (A low concentration of MANY senders is suspicious)
- Sender Diversity Ratio: Unique senders / total received trx (lots of different counterparties sending money is suspicious)
- Received-to-send ratio: Amount received/ amount send (receiving more way more than sending - fan-in indicator)

In [4]:
def compute_fanin_features(df_transactions, snapshot, window_start, prefix):
    """
    Compute fan-in features for ALL accounts within a time window.
    """
    window_trx = df_transactions[(df_transactions["date"] >= window_start) & (df_transactions["date"] < snapshot)]

    # 1. Unique senders per receiver
    unique_senders = window_trx.groupby("RECEIVER_ACCOUNT_ID")[
        "SENDER_ACCOUNT_ID"
    ].nunique()

    # 2. Received volume concentration — % of total received amount from top sender
    amt_per_sender = window_trx.groupby(["RECEIVER_ACCOUNT_ID", "SENDER_ACCOUNT_ID"])[
        "TX_AMOUNT"
    ].sum()
    received_concentration = amt_per_sender.groupby(level=0).apply(
        lambda x: x.max() / x.sum()
    )

    # 3. Sender diversity — unique senders / total received transactions
    trx_received = window_trx.groupby("RECEIVER_ACCOUNT_ID")["TX_ID"].count()
    sender_diversity = unique_senders / trx_received

    # 4. Received to sent ratio
    total_received = window_trx.groupby("RECEIVER_ACCOUNT_ID")["TX_AMOUNT"].sum()
    total_sent = window_trx.groupby("SENDER_ACCOUNT_ID")["TX_AMOUNT"].sum()
    total_received.index.name = "ACCOUNT_ID"
    total_sent.index.name = "ACCOUNT_ID"
    recv_sent_ratio = total_received / total_sent.reindex(
        total_received.index, fill_value=1
    ).clip(lower=1)

    # Combine 
    fanin = pd.DataFrame(
        {
            f"{prefix}_unique_senders": unique_senders,
            f"{prefix}_received_concentration": received_concentration,
            f"{prefix}_sender_diversity": sender_diversity,
            f"{prefix}_recv_sent_ratio": recv_sent_ratio,
        }
    )
    fanin.index.name = "ACCOUNT_ID"

    return fanin

# Test on April 1 snapshot, recent window (March)
test_snapshot = snapshot_dates[0]
test_recent_start = test_snapshot - pd.DateOffset(months=1)

fanin_features = compute_fanin_features(df_transactions, test_snapshot, test_recent_start, "recent")
print(f"\nSnapshot: {test_snapshot.date()}")
print(f"\nWindow: {test_recent_start.date()} to {test_snapshot.date()}")
print(f"\nAccounts with fan-in features: {len(fanin_features)}")
print(f"\nSample (first 5):")
fanin_features.head()



Snapshot: 2020-04-01

Window: 2020-03-01 to 2020-04-01

Accounts with fan-in features: 9778

Sample (first 5):


,recent_unique_senders,recent_received_concentration,recent_sender_diversity,recent_recv_sent_ratio
ACCOUNT_ID,,,,
3,1,1.00,0.33,0.16
5,1,1.00,0.50,0.12
6,1,1.00,0.33,0.80
7,1,1.00,0.33,0.65
10,1,1.00,0.33,0.82


## 4.c Cycle Signals

Features that capture circular flow patters (money moving in loops between accounts). From the EDA, we found that in ~95% of cycle alerts, senders become receivers for the same alert id.

**Why recent window only (no baseline)?** Unlike intensity or fan-in, the presence of bidirectional relationships is the signal itself. An account with 10 bidirectional partners is suspicious whether that's an increase from 5 or has always been 10. The velocity features already capture behavioral change through intensity and fan-in metrics. Cycle velocity could be added as a future improvement.

Features:
- Bidirectional parties 
- Sender-receiver-overlap: % of counterparties that are both senders to and receivers from this client


In [5]:
def compute_cycle_features(df_transactions, snapshot, window_start, prefix):

    """
    compute cycle features for all accounts wihin a given time window
    """

    # trx of interest
    window_trx = df_transactions[(df_transactions['date'] >= window_start) & (df_transactions['date']< snapshot)]

    # For each account find the account_ids the client is sending money to
    sent_to = window_trx.groupby('SENDER_ACCOUNT_ID')['RECEIVER_ACCOUNT_ID'].apply(set) 
    sent_to.index.name = "ACCOUNT_ID"

    # For each account find the account_ids they are receiving money from
    received_from = window_trx.groupby('RECEIVER_ACCOUNT_ID')['SENDER_ACCOUNT_ID'].apply(set)
    received_from.index.name = "ACCOUNT_ID"

    # To ensure sent to and received from have the same index (in case an account didn't receive/send money)
    all_accounts = sent_to.index.union(received_from.index)
    sent_to = sent_to.reindex(all_accounts, fill_value=set())
    received_from = received_from.reindex(all_accounts, fill_value=set())

    # 1. Find bidirectional partners: # of accounts a client is both sending money and receiving money from
    bidirectional = sent_to.combine(received_from, lambda s,r:len(s&r))

    # 2. Overlap ratio: # cps with bidirectional relationship / # total counterparties
    total_counterparties = sent_to.combine(received_from, lambda s,r: (len(s|r)))
    overlap_ratio = bidirectional / total_counterparties.replace(0,1) # in case no cps - replace it with 1 to avoid err

    # Combine
    cycle = pd.DataFrame({

        f"{prefix}_bidirectional_partners": bidirectional,
        f"{prefix}_overlap_ratio":overlap_ratio})
    
    cycle.index.name = 'ACCOUNT_ID'

    return cycle

# Test on April 1 snapshot, recent window (March)
test_snapshot = snapshot_dates[0]
test_recent_start = test_snapshot - pd.DateOffset(months=1)

cycle_features = compute_cycle_features(df_transactions, test_snapshot, test_recent_start, "recent")
print(f"Snapshot: {test_snapshot.date()}")
print(f"Window: {test_recent_start.date()} to {test_snapshot.date()}")
print(f"Accounts with cycle features: {len(cycle_features)}")
print(f"\nSample (first 5):")
print(cycle_features.head())
print(f"\nAccounts with any bidirectional partners: {(cycle_features['recent_bidirectional_partners'] > 0).sum()}")    
print(cycle_features['recent_bidirectional_partners'].describe())


Snapshot: 2020-04-01
Window: 2020-03-01 to 2020-04-01
Accounts with cycle features: 9995

Sample (first 5):
            recent_bidirectional_partners  recent_overlap_ratio
ACCOUNT_ID                                                     
1                                       0                  0.00
2                                       0                  0.00
3                                       0                  0.00
4                                       0                  0.00
5                                       0                  0.00

Accounts with any bidirectional partners: 1716
count   9,995.00
mean        0.20
std         0.51
min         0.00
25%         0.00
50%         0.00
75%         0.00
max         7.00
Name: recent_bidirectional_partners, dtype: float64


## 4.d Velocity/ Change

Velocity features compare recent (previous 1 month) to baseline (2 months before that) to detect behavioral shifts.
A sudden change in behavior is often more suspicious than absolute level of activity.

Features:

- `velocity_trx_sent` — recent sent count / baseline avg monthly sent count
- `velocity_trx_received` — recent received count / baseline avg monthly received count
- `velocity_amount_sent` — recent sent amount / baseline avg monthly sent amount
- `velocity_amount_received` — recent received amount / baseline avg monthly received amount
- `velocity_unique_senders` — recent unique senders / baseline avg unique senders
- `new_counterparty_ratio` — % of counterparties in recent window not seen in baseline

In [6]:
def compute_velocity_features(recent_intensity, baseline_intensity, recent_fanin, baseline_fanin):

    """
    Compute velocity features by comparing recent and baseline features

    Baseline is divided by 2 to get the avg before the comparison
    """

    idxx = recent_intensity.index
    velocity = pd.DataFrame(index = idxx)

    # 1. Baseline monthly averages 
    bl_avg_trx_sent = baseline_intensity['baseline_trx_sent'] / 2
    bl_avg_trx_received = baseline_intensity['baseline_trx_received'] / 2
    bl_avg_amt_sent = baseline_intensity['baseline_amt_sent'] / 2
    bl_avg_amt_received = baseline_intensity['baseline_amt_received'] / 2
    bl_avg_unique_senders = baseline_fanin['baseline_unique_senders']/2

    # 2. Compute Velocity Ratios ---- reindex to ensure same index, replace 0s with 1 to avoid division err
    velocity["velocity_trx_sent"] = recent_intensity["recent_trx_sent"] / bl_avg_trx_sent.reindex(idxx, fill_value=1).clip(lower=1)
    velocity["velocity_trx_received"] = recent_intensity["recent_trx_received"] / bl_avg_trx_received.reindex(idxx, fill_value=1).clip(lower=1)
    velocity["velocity_amt_sent"] = recent_intensity["recent_amt_sent"] / bl_avg_amt_sent.reindex(idxx, fill_value=1).clip(lower=1)
    velocity["velocity_amt_received"] = recent_intensity["recent_amt_received"] / bl_avg_amt_received.reindex(idxx, fill_value=1).clip(lower=1)
    velocity["velocity_unique_senders"] = recent_fanin["recent_unique_senders"] / bl_avg_unique_senders.reindex(idxx, fill_value=1).clip(lower=1)

    velocity.index.name = "ACCOUNT_ID"
    return velocity.fillna(0)

### Velocity Test ###
test_snapshot = snapshot_dates[0]
test_recent_start= test_snapshot - pd.DateOffset(months = 1)
test_baseline_start = test_snapshot - pd.DateOffset(months = 3)

# Compute recent and baseline features first
recent_intensity = compute_intensity_features(df_transactions, test_snapshot, test_recent_start, "recent")
baseline_intensity = compute_intensity_features(df_transactions, test_recent_start, test_baseline_start, "baseline")
recent_fanin = compute_fanin_features(df_transactions, test_snapshot, test_recent_start, "recent")
baseline_fanin = compute_fanin_features(df_transactions, test_recent_start, test_baseline_start, "baseline")

# Compute Velocity Features
velocity_features_test = compute_velocity_features(recent_intensity,baseline_intensity, recent_fanin, baseline_fanin)
print(f"Snapshot: {test_snapshot.date()}")
print(f"Recent: {test_recent_start.date()} to {test_snapshot.date()}")
print(f"Baseline: {test_baseline_start.date()} to {test_recent_start.date()}")
print(f"Accounts: {len(velocity_features_test)}")
print(f"\nSample (first 5):")
print(velocity_features_test.describe())


Snapshot: 2020-04-01
Recent: 2020-03-01 to 2020-04-01
Baseline: 2020-01-01 to 2020-03-01
Accounts: 9995

Sample (first 5):
       velocity_trx_sent  velocity_trx_received  velocity_amt_sent  velocity_amt_received  velocity_unique_senders
count           9,995.00               9,995.00           9,995.00               9,995.00                 9,995.00
mean                1.20                   1.05             130.06               1,215.94                     1.79
std                 1.65                   0.34           9,109.57              23,163.67                     0.58
min                 0.00                   0.00               0.00                   0.00                     0.00
25%                 0.89                   0.92               1.00                   0.71                     1.50
50%                 1.00                   1.03               1.04                   1.06                     2.00
75%                 1.25                   1.17               2.00      

## 4.e New Counterparty Ratio 

What % of counterparties in the recent window are new (not seen in the baseline window).

A high ratio means client is suddenly transaction with lots of new people, which is suspicious for both fan-in and cycle patterns.

In [7]:
def compute_new_counterparty_ratio(df_transactions, snapshot, recent_start, baseline_start):

    """
    Compute % of counterparties in recent window, which were not seen in baseline window.
    """

    recent_trx = df_transactions[(df_transactions['date'] >= recent_start) & (df_transactions['date'] < snapshot)]
    baseline_trx = df_transactions[(df_transactions['date'] >= baseline_start) & (df_transactions['date'] < recent_start)]

    # get counterparties per account per window
    def get_counterparties(trx, account_col, counterparty_col):
        return trx.groupby(account_col)[counterparty_col].apply(set).rename_axis('ACCOUNT_ID')
    
    recent_sent = get_counterparties(recent_trx, 'SENDER_ACCOUNT_ID', 'RECEIVER_ACCOUNT_ID')
    recent_received = get_counterparties(recent_trx, 'RECEIVER_ACCOUNT_ID', 'SENDER_ACCOUNT_ID')
    baseline_sent = get_counterparties(baseline_trx, 'SENDER_ACCOUNT_ID', 'RECEIVER_ACCOUNT_ID')
    baseline_received = get_counterparties(baseline_trx, 'RECEIVER_ACCOUNT_ID', 'SENDER_ACCOUNT_ID')

    # all accounts with any trx made in the recent window (reindexing with recent window is sufficient)
    all_accounts = recent_sent.index.union(recent_received.index)

    # reindex sent and received to have to complete list of clientids in the set - if nothing sent/received, fill with set()
    recent_sent_filled = recent_sent.reindex(all_accounts, fill_value = set())
    recent_received_filled = recent_received.reindex(all_accounts, fill_value = set())
    recent_all = recent_sent_filled.combine(recent_received_filled, lambda s,r: s|r)
    
    baseline_sent_filled = baseline_sent.reindex(all_accounts, fill_value = set())
    baseline_received_filled = baseline_received.reindex(all_accounts, fill_value = set())
    baseline_all = baseline_sent_filled.combine(baseline_received_filled, lambda s,r: s|r)

    # New counterparty ratio
    new_ratio = recent_all.combine(baseline_all, lambda recent, base: len(recent-base)/len(recent) if len(recent)>0 else 0)

    return pd.DataFrame({"new_counterparty_ratio": new_ratio})


test_snapshot = snapshot_dates[0]
test_recent_start = test_snapshot - pd.DateOffset(months=1)
test_baseline_start = test_snapshot - pd.DateOffset(months=3)

new_cp = compute_new_counterparty_ratio(df_transactions, test_snapshot, test_recent_start, test_baseline_start)
print(f"Accounts: {len(new_cp)}")
print(f"\nSample:")
print(new_cp.head(10))
print(f"\nMean new counterparty ratio: {new_cp['new_counterparty_ratio'].mean():.2f}")
print(f"Max new counterparty ratio: {new_cp['new_counterparty_ratio'].max():.2f}")


Accounts: 9995

Sample:
            new_counterparty_ratio
ACCOUNT_ID                        
1                             0.00
2                             0.00
3                             0.00
4                             0.00
5                             0.00
6                             0.00
7                             0.00
8                             0.00
9                             0.00
10                            0.00

Mean new counterparty ratio: 0.09
Max new counterparty ratio: 1.00


# 5) Forward-Looking Labels

For each client at each snapshot date, the label answers: **will this client be involved in suspicious activity during the upcoming calendar month?**

- April 1 snapshot → label = 1 if client has alert in April
- May 1 snapshot → label = 1 if client has alert in May
- June 1 snapshot → label = 1 if client has alert in June

**Label logic:** An account is flagged (label = 1) only if BOTH conditions are met:
1. Involved in an alert transaction during the snapshot month (as sender or receiver)
2. Marked as a fraud account (`IS_FRAUD = 1`) in the accounts table

This dual condition ensures we only label accounts that are genuinely part of the laundering scheme.
The `IS_FRAUD` flag is our ground truth, regardless of whether the account acts as sender or receiver in the alert.

**Why alerts instead of IS_FRAUD alone?** The static `IS_FRAUD` flag doesn't tell us WHEN fraud happened. Using alerts gives us temporal labels, the same fraud account can be labeled 1 in April but 0 in May, letting the model learn what behavioral patterns PRECEDE suspicious activity. This is an approximation of True Positive labels for alerts.


**Observation:** In AMLSim, every account involved in an alert is also marked as `IS_FRAUD = 1`, there are zero false positives in the synthetic alert system. This means the dual condition (alerts ∩ IS_FRAUD) simplifies to just using alerts for this dataset. However, keeping the dual logic is good practice because in production AML systems, alerts frequently include FP accounts caught in the net.

In [8]:
def compute_labels(df_alerts, df_accounts, snapshot):
    """
    Label = 1 if account is fraud and appears in an alert during snapshot month.
    """

    month_start = snapshot
    month_end = snapshot + pd.offsets.MonthEnd(0)

    # Alerts in snapshot month
    window_alerts = df_alerts[
        (df_alerts["date"] >= month_start) & 
        (df_alerts["date"] <= month_end)
    ]
    
    # accounts involved in alerted transactions
    alerted_accounts = set(window_alerts['SENDER_ACCOUNT_ID']) | set(window_alerts['RECEIVER_ACCOUNT_ID'])

    # Only label accounts that are actually fraud (TP Appr.)
    fraudulent_accounts = set(df_accounts[df_accounts['IS_FRAUD'] == 1]['ACCOUNT_ID'])
    labeled_accounts = alerted_accounts & fraudulent_accounts

    print(f" On {snapshot.date()}, there were {len(window_alerts)} alerted transactions, {len(alerted_accounts)} accounts involved in an alerted trx and {len(labeled_accounts)} accounts with confirmed Fraud")
    return labeled_accounts

# Test on all snapshots
for snapshot in snapshot_dates:
    flagged = compute_labels(df_alerts, df_accounts, snapshot)


 On 2020-04-01, there were 200 alerted transactions, 235 accounts involved in an alerted trx and 235 accounts with confirmed Fraud
 On 2020-05-01, there were 246 alerted transactions, 292 accounts involved in an alerted trx and 292 accounts with confirmed Fraud
 On 2020-06-01, there were 293 alerted transactions, 342 accounts involved in an alerted trx and 342 accounts with confirmed Fraud


# 6) Final Dataset Construction

Loop through all 3 snapshots, compute all features and labels, and combine them into a single dataset.

For each snapshot:

1. Compute recent (snapshot - 1m) and baseline features (recent - 2m)
2. Compute velocity features: Comparing recent vs.baseline
3. Compute new counterparty ratio. 
4. Compute forward looking labels based on alerts
5. Join everything into a one row per client.

In [9]:
all_snapshots = []

for snapshot in snapshot_dates:
    print(f"\n{'*'*60}\nProcessing snapshot: {snapshot.date()}\n{'*'*60}")

    # define windows
    recent_start = snapshot - pd.DateOffset(months = 1)
    baseline_start = snapshot - pd.DateOffset(months = 3)

    print(f" Recent: {recent_start} to {snapshot.date()}")
    print(f" Baseline: {baseline_start} to {recent_start.date()}")


    # 1. Intensity features
    recent_intensity = compute_intensity_features(df_transactions, snapshot, recent_start, "recent")
    baseline_intensity = compute_intensity_features(df_transactions, recent_start, baseline_start, "baseline")

    # 2. Fan-in Features
    recent_fanin = compute_fanin_features(df_transactions, snapshot, recent_start, "recent")
    baseline_fanin = compute_fanin_features(df_transactions, recent_start, baseline_start, "baseline")

    # 3. Cycle Features - recent only 
    recent_cycle = compute_cycle_features(df_transactions, snapshot, recent_start, "recent")

    # 4. Velocity features
    velocity = compute_velocity_features(recent_intensity, baseline_intensity, recent_fanin, baseline_fanin)
    
    # 5. New counterparty ratio
    new_cp = compute_new_counterparty_ratio(df_transactions, snapshot, recent_start, baseline_start)
    
    # 6. Labels
    flagged_accounts = compute_labels(df_alerts, df_accounts, snapshot)


    # 7. Join all features
    snapshot_df = recent_intensity \
        .join(baseline_intensity, how="outer") \
        .join(recent_fanin, how="outer") \
        .join(baseline_fanin, how="outer") \
        .join(recent_cycle, how="outer") \
        .join(velocity, how="outer") \
        .join(new_cp, how="outer") \
        .fillna(0)
    

    # 8. Add metadata 
    snapshot_df['snapshot_date'] = snapshot
    snapshot_df['label'] = snapshot_df.index.isin(flagged_accounts).astype(int)

    print(f"Accounts: {len(snapshot_df)}")
    print(f" Number of accounts flagged: {snapshot_df['label'].sum()}")

    # combine dfs for each snapshot    
    all_snapshots.append(snapshot_df)

# combine list of dfs into a single df
modeling_df = pd.concat(all_snapshots).reset_index()
print(f"\n{'='*60}")
print(f"Final dataset: {modeling_df.shape[0]} rows × {modeling_df.shape[1]} columns")
print(f"Label distribution:\n{modeling_df['label'].value_counts()}")



************************************************************
Processing snapshot: 2020-04-01
************************************************************
 Recent: 2020-03-01 00:00:00 to 2020-04-01
 Baseline: 2020-01-01 00:00:00 to 2020-03-01
 On 2020-04-01, there were 200 alerted transactions, 235 accounts involved in an alerted trx and 235 accounts with confirmed Fraud
Accounts: 9999
 Number of accounts flagged: 235

************************************************************
Processing snapshot: 2020-05-01
************************************************************
 Recent: 2020-04-01 00:00:00 to 2020-05-01
 Baseline: 2020-02-01 00:00:00 to 2020-04-01
 On 2020-05-01, there were 246 alerted transactions, 292 accounts involved in an alerted trx and 292 accounts with confirmed Fraud
Accounts: 9999
 Number of accounts flagged: 292

************************************************************
Processing snapshot: 2020-06-01
************************************************************
 

# 7) Dataset Validation


Before starting with the modeling, let's verify that dataset is clean and free of data leakage.

In [10]:
# 1. Nulls and infinites

print("*** Null Check ***")
null_counts = modeling_df.isnull().sum()
if null_counts.sum() == 0:
    print('\nNo null values')
else:
    print('\nNulls found')
    print(null_counts[null_counts > 0])

print("\n*** Infinity Check ***")
numeric_cols = modeling_df.select_dtypes(include = [np.number]).columns
inf_counts = np.isinf(modeling_df[numeric_cols]).sum()
if inf_counts.sum() == 0:
    print("\n✅ No infinite values")
else:
    print("❌ Infinities found:")
    print(inf_counts[inf_counts > 0])


# 2. Class balance per snapshot
print("\n*** Class balance ***")
for snapshot in snapshot_dates:
    snapshot_data = modeling_df[modeling_df['snapshot_date'] == snapshot]
    fraud_count = snapshot_data['label'].sum()
    total = snapshot_data.shape[0]
    print(f"  {snapshot.date()} → {fraud_count} fraud / {total} total ({fraud_count/total*100:.2f}%)")

print(f'\nClass distribution across all snapshots: {modeling_df["label"].mean() * 100:.2f}%')

# 3. Feature Summary

print("\n*** Feature Summary ***")

feature_cols = [c for c in modeling_df.columns if c not in ["ACCOUNT_ID", "snapshot_date", "label"]]
print(f"\nTotal features: {len(feature_cols)}")
print(f"\nFeature Stats: ")
print()
print(modeling_df[feature_cols].describe().T[["mean", "std", "min", "max"]].to_string())


*** Null Check ***

No null values

*** Infinity Check ***

✅ No infinite values

*** Class balance ***
  2020-04-01 → 235 fraud / 9999 total (2.35%)
  2020-05-01 → 292 fraud / 9999 total (2.92%)
  2020-06-01 → 342 fraud / 9999 total (3.42%)

Class distribution across all snapshots: 2.90%

*** Feature Summary ***

Total features: 28

Feature Stats: 

                                      mean          std  min           max
recent_trx_sent                      20.35        21.67 0.00        198.00
recent_amt_sent                 758,775.00 3,107,406.65 0.00 76,225,345.00
recent_avg_amt_sent             232,680.61   870,800.89 0.00 19,463,548.00
recent_trx_received                  20.35        29.94 0.00        762.00
recent_amt_received             758,775.00 2,218,131.72 0.00 57,577,116.17
recent_avg_amt_received          65,541.69   270,011.72 0.00 12,154,227.00
baseline_trx_sent                    39.84        40.93 0.00        337.00
baseline_amt_sent               226,062.09   86

In [11]:
print(modeling_df["velocity_amt_received"].describe(percentiles=[0.9, 0.95, 0.99]))
print()
print(modeling_df["velocity_amt_sent"].describe(percentiles=[0.9, 0.95, 0.99]))


count      29,997.00
mean        5,272.23
std       125,765.54
min             0.00
90%           109.35
95%           658.61
99%        11,400.71
max     7,743,013.87
Name: velocity_amt_received, dtype: float64

count       29,997.00
mean         1,593.59
std        126,998.86
min              0.00
90%              7.20
95%             12.88
99%             35.86
max     15,515,128.37
Name: velocity_amt_sent, dtype: float64


### Handling Extreme Values in Velocity Features

Looking at the 99th percentile vs the max:

| Feature | 99th percentile | Max |
|---|---|---|
| velocity_amt_sent | 35.86 | 15,515,128 |
| velocity_amt_received | 11,400 | 7,743,013 |

99% of accounts have reasonable velocity values, but a few extreme outliers are inflating the statistics dramatically.

**Root cause:** Division by near-zero baseline values. When an account had tiny activity in the baseline period (e.g., €0.50 sent) and normal activity in the recent period (e.g., €1,000), the velocity ratio becomes huge (2,000) — even though it's technically a "real" behavioral change, the magnitude is misleading.

**Why this matters for modeling:**
- **Logistic Regression** is sensitive to feature scale. A feature ranging 0 to 15M would dominate every other feature (which mostly range 0-100), making the model ignore smaller but meaningful signals.
- **LightGBM** handles extremes fine (it splits on thresholds), but we want consistent preprocessing across models.

**Fix applied:** Log-transform (`np.log1p`) to compress the scale while preserving **ordering**. This turns a max of 15.5M into ~16.6, bringing it in line with other features. Relative differences are preserved a 100x change is still twice as extreme as a 10x change after the transform.

We apply this **after** the initial dataset construction and only to the velocity features that showed extreme outliers, to make the sanity check and reasoning visible in the notebook rather than hiding the fix inside the feature function.

In [12]:
cols_to_transform = [
    "velocity_amt_sent",
    "velocity_amt_received",
    "velocity_trx_sent",
    "velocity_trx_received",
    "recent_recv_sent_ratio",
    "baseline_recv_sent_ratio",
]
for col in cols_to_transform:
    modeling_df[col] = np.log1p(modeling_df[col])

# Verify the transform
print("After log-transform:")
print(modeling_df[cols_to_transform].describe(percentiles=[0.9, 0.95, 0.99]))


After log-transform:
       velocity_amt_sent  velocity_amt_received  velocity_trx_sent  velocity_trx_received  recent_recv_sent_ratio  baseline_recv_sent_ratio
count          29,997.00              29,997.00          29,997.00              29,997.00               29,997.00                 29,997.00
mean                1.00                   1.72               0.72                   0.69                    2.41                      1.91
std                 0.78                   2.19               0.28                   0.16                    2.94                      2.05
min                 0.00                   0.00               0.00                   0.00                    0.00                      0.00
90%                 2.10                   4.70               1.10                   0.85                    6.96                      5.09
95%                 2.63                   6.49               1.25                   0.92                    8.14                      6.05

### Diagnostic: Cycle Feature Discriminative Power

Let's verify whether cycle features actually separate fraud from non-fraud accounts. A feature with no signal adds noise without helping the model.
**Finding:** Cycle features show weak discriminative signal — fraud accounts have only ~30% more bidirectional partners on average (0.26 vs 0.20). This suggests cycle patterns in the 1-month window rarely complete within the snapshot period. Fan-in features are expected to carry the main detection signal. Future work could extend cycle detection across multiple snapshot windows.

In [13]:
apr_snap = modeling_df[modeling_df["snapshot_date"] == snapshot_dates[0]]
print("Non-fraud bidirectional partners:")
print(apr_snap[apr_snap["label"] == 0]["recent_bidirectional_partners"].describe())
print("\nFraud bidirectional partners:")
print(apr_snap[apr_snap["label"] == 1]["recent_bidirectional_partners"].describe())


Non-fraud bidirectional partners:
count   9,764.00
mean        0.20
std         0.50
min         0.00
25%         0.00
50%         0.00
75%         0.00
max         7.00
Name: recent_bidirectional_partners, dtype: float64

Fraud bidirectional partners:
count   235.00
mean      0.26
std       0.68
min       0.00
25%       0.00
50%       0.00
75%       0.00
max       5.00
Name: recent_bidirectional_partners, dtype: float64


# 8. Export

In [14]:
# save the modeling df
output_path = Path("../data/processed")
output_path.mkdir(parents = True, exist_ok = True)

modeling_df.to_csv(output_path / "modeling_dataset.csv", index=False)

print(modeling_df.shape)


(29997, 31)


# 9) Summary & Next Steps

## Dataset Overview
- **Shape:** 29,997 rows × 31 columns (9,999 accounts × 3 monthly snapshots)
- **Features:** 28 behavioral features across 5 categories
- **Label:** ~2.9% positive rate (fraud accounts with alerts in the snapshot month)

## Feature Categories
| Category | Count | Description |
|---|---|---|
| Intensity (recent) | 6 | Transaction counts, amounts, averages (sent/received) |
| Intensity (baseline) | 6 | Same metrics for baseline period |
| Fan-in (recent) | 4 | Unique senders, concentration, diversity, recv/sent ratio (log-transformed) |
| Fan-in (baseline) | 4 | Same metrics for baseline period (recv/sent ratio log-transformed) |
| Cycle | 2 | Bidirectional partners, overlap ratio |
| Velocity | 5 | Recent vs baseline ratios (log-transformed for LR compatibility) |
| New counterparty | 1 | % of counterparties not seen in baseline |

## Validation Findings & Fixes Applied

**1. Extreme outliers in scale-sensitive features**
- `velocity_amt_sent` had max of 15.5M, `velocity_amt_received` had max of 7.7M
- `recent_recv_sent_ratio` had max of 16.8M, `baseline_recv_sent_ratio` had max of 1.8M
- **Root cause:** Division by near-zero denominators when baseline activity was tiny. When an account had minimal baseline activity (e.g., €0.50 sent) and normal recent activity (e.g., €1,000), the ratio became huge even though the behavioral change was modest.
- **Fix 1 (in velocity function):** Added `.clip(lower=1)` to baseline denominators to prevent division-by-near-zero artifacts at the source.
- **Fix 2 (after dataset creation):** Applied `np.log1p()` to velocity features AND recv/sent ratio features to compress remaining extreme tails. This preserves ordering (bigger is still bigger) while bringing all scale-sensitive features onto a comparable range for Logistic Regression.

**2. Cycle features show weak discriminative signal**
- After fixing a bug in the cycle computation (see Section 4.c), `bidirectional_partners` showed only ~30% difference between fraud (mean 0.26) and non-fraud (mean 0.20) accounts.
- This suggests cycle patterns in the 1-month snapshot window rarely complete within the observation period — money often flows through the circle across multiple months.
- Fan-in features are expected to carry the main detection signal.
- **Future improvement:** Extend cycle detection across multiple snapshot windows, or add amount-matching features (per EDA finding that cycle alerts share a constant transaction amount).

**3. Class imbalance (~2.9% positive rate)**
- Healthier than the 0.13% at transaction level, but still imbalanced.
- Will need class weights, threshold tuning, and Precision@TopK as the evaluation metric during modeling.

**4. AMLSim observation: zero false positives**
- All accounts involved in alerts are also `IS_FRAUD = 1`. In production AML systems, alerts frequently include innocent accounts caught in the net.
- Dual labeling logic (alerts ∩ IS_FRAUD) is kept for production-readiness even though it simplifies to alerts alone for this synthetic dataset.

## Next Steps
→ See `03_modeling.ipynb`:
1. **Time-based train/test split:** April + May snapshots for training, June for testing
2. **Rule-based baseline:** Simple threshold rules on fan-in features (e.g., flag if unique_senders > threshold)
3. **Logistic Regression:** Interpretable risk scoring model with class weights and feature scaling
4. **LightGBM:** Non-linear model to capture feature interactions
5. **Evaluation:** Precision@TopK, recall within alert budget, SHAP explainability